realtek: add support for DGS-1210-52 (relaunch)

plappermaul opened the pull request:
Since a few weks PR #10227 is open to provide support for the DGS-1210-52 device. In between the D-Link device tree was rebuilt and the PR has not yet been adapted to that. Compiling the clock/timer drivers all the day long with not yet published fixes I relaunch the support request with a new PR.

Thanks to @luizluca who provided the initial PR and the work to identify the GPIOs. In case the contribution should be reflected in the commits do not hesitate to request changes.

plappermaul commented:
Developing each day with an DGS-1210-52 I really miss this in the repository.

luizluca commented:
I might review and test this today. Although I do not have the required commit rights, it'll make it easier.

plappermaul commented:
> I might review and test this today. Although I do not have the required commit rights, it'll make it easier.

Hi, do whatever you like. E.g. take over the commits and continue with your initial PR.

luizluca commented:
> > I might review and test this today. Although I do not have the required commit rights, it'll make it easier.
> 
> Hi, do whatever you like. E.g. take over the commits and continue with your initial PR.

I guess I wasn't clear. It is not about commit authorship. You can add my "Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>" to that last commit if you want to credit it. I just don't have the privileges to merge this commit.
I tested it and it works as before (r20650+3-8b06e06832). However, I did not see any improvements on two missing points:

1) combo ports are still unavailable
2) tagged vlans that pass through the switch gets the tag removed when it leaves the switch. Tagged traffic directed to the switch does work as expected. 

```
lan47             124 PVID Egress Untagged
                  228
lan48             12
                  124 PVID Egress Untagged
                  214
                  228
                  601
                  669
```

Vlan 228 received from lan48 is received by a device connected to lan47 as untagged traffic.
As this is something related to the SoC and not this specific device, there is no reason for not merging it.

And this branch already needs to be rebased.

plappermaul commented:
> 2\. tagged vlans that pass through the switch gets the tag removed when it leaves the switch. Tagged traffic directed to the switch does work as expected.

I agree that this is not part of this PR. Nevertheless can you tell me if one can reproduce that issue on the normal (non SFP) ports?

luizluca commented:
> > 2. tagged vlans that pass through the switch gets the tag removed when it leaves the switch. Tagged traffic directed to the switch does work as expected.
> 
> I agree that this is not part of this PR. Nevertheless can you tell me if one can reproduce that issue on the normal (non SFP) ports?

Yes. I'm not using SPF ports. Just configure ports lan1 and lan2 with a tagged vlan 10 and untagged vlan 11. Send a tagged traffic (vlan 10) through lan1. Capture lan2 traffic externally. You might see data without the vlan 10 tag (as untagged). Vlan 11, as expected, goes and comes without issues.

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_d-link_dgs-1210-52.dts:
```diff
@@ -0,0 +1,164 @@
+		EXTERNAL_PHY(47)
+
+		/* External phy RTL8214FC? or
+		  2xRTL8393 SERDES + 2xRealtek RTL8218B (external)? */
```
Commit message says "The 4 combo ports were not stable and were commented out," but these ports aren't disabled in this version. Can you open the device to visually verify which chip is on the board? If there's a heat sink that doesn't come off, the RTL8218B has a rectangular footprint, while the RTL8214FC has a square footprint. See e.g. (DGS-1210-16)[https://svanheule.net/switches/dgs-1210-16], which also has both.

It looks like `rtl8218b_ext_match_phy_device()` in `target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c` doesn't understand yet that an RTL8214FC can also be used on non-RTL838x platforms, so that needs fixing for this phy to be detected.

While you're making changes, please also run checkpatch.pl and fix any warnings that can be fixed. The less style issues there are, the easier it is to spot actual errors.

plappermaul replied:
DGS-1210-52 uses a RTL8214FC. Commit messages updated to reflect that.

svanheule replied:
Thanks, I think this can be merged the.

plappermaul committed:
realtek: D-Link make common DT include generic

The D-Link DGS device tree was reorganized to better reflect the common
DT parts. The common include is named SOC specific (838X) and it seemed
like a good choice to add another common include in the future for the
RTL839X devices. From the current point of view this option is not really
needed.

1. The common part only includes data that matches RTL839X devices too.
2. The Panasonic DT structure avoids including the basic DTSI inside the
   common DTSI.

Taking simplicity of the Panasonic include logic and in perparation to
provide DGS-1210-52 support it makes sense to harmonize this.

- rename common include to reflect its content
- move the link to the root DTSI directly to the device specific DTS

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: align DT macros in RTL839X with RTL838X

Add a missing definition to the RTL839X DT.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: add support for DGS-1210-52

Hardware specification
----------------------

* RTL8393M SoC, 1 MIPS 34Kc core @ 700MHz
* 128MB DRAM
* 32MB NOR Flash
* 48 x 10/100/1000BASE-T ports
  - 6 x External PHY with 8 ports (RTL8218D)
* 4 x Gigabit RJ45/SFP Combo ports
  - External PHY with 4 SFP ports (RTL8214FC)
* Power LED
* Reset button on front panel
* UART (115200 8N1) via unpopulated standard 0.1" pin header marked J14

The gpio-restart node is not required but it does reset the switch.

TODO: The 4 combo ports attached to the RTL8314FC are not detect
properly. Linux kernel reports 49 and 50 as "External RTL8393 SERDES"
and 51 and 52 as "RTL8218B (external)". Those ports only work if
u-boot initialize it (for example, loading initramfs image using one
of those ports). A patch to PHY detection is needed for full support.

The firmware recovery using U-Boot is broken for all DGS-1210 tested
devices as pressing RESET does not trigger it (only if pressed from a
running stock image)

UART pinout
-----------

[o]ooo|J14
 | ||`------ GND
 | |`------- RX
 | `-------- TX
 `---------- Vcc (3V3)

Installation using OEM upgrade
------------------------------

1. Make sure you are running OEM firmware in image2 slot (logged as admin):
   - > config firmware image_id 2 boot_up
   - > reboot
2. Install squashfs-factory_image1.bin to image1 using (logged as admin):
   - > download firmware_fromTFTP <tftpserver> factory_image1.bin
   - > config firmware image_id 1 boot_up
   - > reboot

Installation using serial interface
-----------------------------------

1. Press Escape key during `Hit Esc key to stop autoboot` prompt
2. Press CTRL+C keys to get into real U-Boot prompt
3. Init network with `rtk network on` command
4. Load image with `tftpboot 0x8f000000 openwrt-realtek-rtl839x-d-link_dgs-1210-52-initramfs-kernel.bin` command
5. Boot the image with `bootm` command

Once booted the initramfs, install the squashfs-sysupgrade.bin as a
normal OpenWrt system.

Dual-boot with stock firmware using writable u-boot-env
-------------------------------------------------------

From stock to OpenWrt / boot image 1 (CLI as admin):
   - > config firmware image_id 1 boot_up
   - > reboot

From OpenWrt to stock / boot image 2: (shell as root)
   - # fw_setenv bootcmd 'run addargs ; bootm 0xb4e80000'
   - # fw_setenv image '/dev/mtdblock7'
   - # reboot

Debrick using serial interface
------------------------------

1. Press Escape key during `Hit Esc key to stop autoboot` prompt
2. In a Windows PC, run 'D-Link Network Assistant v2.0.2.4'. It should
   detect the switch
3. Flash the firmware.

Back to stock firmware using dual-boot
--------------------------------------

If you have serial interface, you can change u-boot env vars
interrupting the boot process. If not but you are running OpenWrt, you
can dual-boot (as mentioned eariler) and skip to step 4:

1. Press Escape key during `Hit Esc key to stop autoboot` prompt
2. Press CTRL+C keys to get into real U-Boot prompt
3. Boot the image 2:
   - set image /dev/mtdblock7; run addargs; bootm 0xb4e80000
4. Once booted, log as admin and change the boot image to 2
   - > config firmware image_id 2 boot_up
   - > reboot
5. After the boot, flash image1 with the vendor image

Back to stock firmware using DNA
--------------------------------

1. From an OpenWrt:
   - # fw_setenv bootstop on
   - # reboot
2. In a Windows PC, run 'D-Link Network Assistant v2.0.2.4'. It should
   detect the switch
3. Flash the firmware.

It has been developed and tested on device with F3 revision.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

svanheule commented:
For the record:
```
 21:08:45 <plappermaul> [  148.240000] sfp sfp-p49: module D-LINK           DEM-311GT        rev 1.0  sn SA0M1IA002289    dc 181022
 21:08:45 <plappermaul> [  161.230000] sfp sfp-p49: module removed
 21:08:46 <plappermaul> [  168.100000] sfp sfp-p50: module D-LINK           DEM-311GT        rev 1.0  sn SA0M1IA002289    dc 181022
 21:08:46 <plappermaul> [  175.160000] sfp sfp-p50: module removed
 21:08:47 <plappermaul> [  182.260000] sfp sfp-p52: module D-LINK           DEM-311GT        rev 1.0  sn SA0M1IA002289    dc 181022
 21:08:47 <plappermaul> [  185.480000] sfp sfp-p52: module removed
 21:08:48 <plappermaul> [  189.930000] sfp sfp-p51: module D-LINK           DEM-311GT        rev 1.0  sn SA0M1IA002289    dc 181022
 21:08:48 <plappermaul> [  200.510000] sfp sfp-p51: module removed
```
```
 21:10:12 <plappermaul> [  308.600000] sfp sfp-p51: module CISCO-PROLABS    GLC-SX-MMD       rev A    sn PGH00059         dc 120712
 21:10:12 <plappermaul> [  315.440000] sfp sfp-p51: module removed
 21:10:13 <plappermaul> [  320.370000] sfp sfp-p52: module CISCO-PROLABS    GLC-SX-MMD       rev A    sn PGH00059         dc 120712
 21:10:13 <plappermaul> [  324.570000] sfp sfp-p52: module removed
 21:10:14 <plappermaul> [  328.140000] sfp sfp-p49: module CISCO-PROLABS    GLC-SX-MMD       rev A    sn PGH00059         dc 120712
 21:10:14 <plappermaul> [  333.540000] sfp sfp-p49: module removed
 21:10:15 <plappermaul> [  336.450000] sfp sfp-p50: module CISCO-PROLABS    GLC-SX-MMD       rev A    sn PGH00059         dc 120712
 21:10:15 <plappermaul> [  339.870000] sfp sfp-p50: module removed
```

Since the SFP modules work, I'll remove the comment in `target/linux/realtek/dts-5.10/rtl839x_d-link_dgs-1210_gpio.dtsi`. Thanks, @plappermaul  and @luizluca!

307e5cfc6ec2 realtek: D-Link make common DT include generic
9ed1a1138eba realtek: align DT macros in RTL839X with RTL838X
e04e821471bb realtek: add support for DGS-1210-52

luizluca commented:
For the record, I found the issue with the vlan tag https://github.com/openwrt/openwrt/pull/11195
